### DISCUSSION TITLE
Copy Loco roster to another throttle.

### DISCUSSION THREAD

**Discussion by Budman1758** [DCC-EX/EX-T3-WiFi#81] [Q&A] [answered]
I am in the process of building a couple more of these and was wondering what the easiest way to replicate the loco roster to another unit would be.

**Comment by alex-code** [comment]
Currently you'd have to remove the MicroSD card to copy it to another.
I need to look into the possibility of zip files on the ESP32, creating and unzipping so you can export and import.

**Reply by Alan-Lomax** [reply]
This sounds like an opportunity for an OTA update feature!. Being able to send a roster as XML or JSON and having the throttle pick it up would be awesome. Personally I would love to be able to export my JMRI roster and send it to the throttle. Even my modest stable of 15 locos means this will be a pain to do by hand. Others I know have a lot more locos than this. Those on the layout are rotated out in 'shifts' after a few weeks / months.

**Reply by alex-code** [reply]
I'm not familiar with JMRI so not sure how to import from that. I guess there'd need to be some conversion to.

I think a bulk/individual export/import would be a good idea.

**Reply by Alan-Lomax** [reply]
I believe the native roster in JMRI is an XML file (I can check later and provide a sample if needed). That might be a distraction though … the key thing is bulk import/export in whatever format. Heck even a csv file to and from excel would be 90% of the problem solved

**Reply by daveclifford** [reply]
If a loco roster were provided to me from JMRI in a file and a sample SD card (zipped up) containing a roster of 2 locos were provided, (like in this messsage thread), I will look into the issue and write up a plan on how a developer would design/develop/test/use a routine to do this off-line.

My plan would be to take the formatted SD card image containing a LOCO file and add the JMRI Loco roster to the existing file, using an off-line system. It appears that a Python routine may offer a common routine to be used by Windows. Mac, and Linux users.

e.g.  loco-roster-merge.py  filename.jmri  filename.sd filename.output

The filename.output file would be created and the user would check the output loco roster file and, if satisfied, replace/rewriite to an new SD card.

Not one push-button execution, but it'll work! More automagic at a later date! 

The routine would prompt for where the SD card file is named/located, where the JMRI  file is named/located and then merge tte JMRI file to the SD Loco roster file. This would be the first version. 

Adding other formats could be done at a later time if there is a demand.... -  but a high priority to me.

It might be painful to use the EX-T3-WiFi throttle to add 12-15 locos, but the ability is currently there.

I feel it is more important for Alex to spend his time fixing/enhancing/adding other functions.

Dave

**Reply by Alan-Lomax** [reply]
Attached in an abbreviated roster showing the format of the native roster file.
More info can be found here: https://www.jmri.org/help/en/html/apps/DecoderPro/RosterFile.shtml

[Sample_JMRI_roster.zip](https://github.com/DCC-EX/EX-T3-WiFi/files/11084792/Sample_JMRI_roster.zip)

**Reply by Alan-Lomax** [reply]
Looking further ... it appears there is a built in (into JMRI) python script to export the roster to CSV   `RosterCsvExport.py`
I have not explored this option but if correct that would be the heavy lifting part of the 'export' done.
Then a matter of conditioning the data (throw most away - keep the bits needed for the throttle) and then coming up with a good way of  dropping that info into the throttle database.

**Reply by Alan-Lomax** [reply]
During a zoom meeting today .. and quite by coincidence .. I saw a good example of using OTA. At the root of the talk was this RNT article: https://randomnerdtutorials.com/esp32-ota-over-the-air-vs-code/
The implementation is only a few lines of code (which initially must be included and uploaded in the conventional manner.) Thereafter pointing a web browser to the ESP and providing login credentials allows you to upload a bin file (your latest code that has been compiled to file). In a few seconds the new code was in and running. It all looked very slick. Apparently the same process can be used to upload other files and so is potentially an opportunity to send to the throttle a file which is the 'roster'. The throttle program changes would involve checking for, reading and parsing such a file)

**Reply by alex-code** [reply]
I'm going to try the browser based upload first as it's a bit more plug'n'play.
I think for OTA you need half the available flash for a new upload to and I'm not sure how much free space there is.

**Reply by daveclifford** [reply]
Alex - 

I agree completely re: space available for buffering. 

Working at IBM, we discovered that automagic updates were cost prohibitive due to the need for 2x+ memory
(and memory was very expensive $100+) , so the solution was to replace eproms - still expensive!

Your solution is elegant and the additional code to add to provide the upload cabability should be trivial.

Installing/using the VS Code and PlatformIO IDE  to build/upload the code deters a lot of model train hobbyists from  building the T3 WiFi Throttle. 

My 3 cents,

Dave

**Comment by Budman1758** [comment]
Is it possible to just expose the SD card as a drive like it is when plugged into your PC? Plug the cable in just as you do when uploading the code and shows up as a drive. Then you can copy config and any other files back and forth. Don't know if this is possible or not. 

Transferring files would probably happen quite a bit more often that code updates I would think. Also during code updates that doesn't really affect the base file system that is uploaded separately from the code when uploading the initial code.

Food for thought.

**Reply by alex-code** [reply]
Unfortunately not, when plugged in I can only communicate via Serial so would need a desktop app too.

I'm nearly done with some download/upload options.
Each loco/function set can be downloaded individually or you can download all configs.
These can then be individually or bulk uploaded.

Just doing testing as too many requests can overwhelm the ESP32 and cause it to crash.

**Comment by alex-code** [comment, marked as answer]
I've updated the repo with the new download/upload options.
You only need to update the filesystem with this one (after downloading the changes).
https://github.com/DCC-EX/EX-T3-WiFi/blob/master/docs/code.md#uploading-the-code

You may need to CTRL+F5 or Shift+F5 in your browser so it downloads the throttle website changes.
